src/compiler/checker.ts

```diff
@@ -2014,6 +2017,34 @@ module ts {
             }
         }
 
+        // Push an entry on the type resolution stack. If an entry with the given target is not already on the stack,
+        // a new entry with that target and an associated result value of true is pushed on the stack, and the value
+        // true is returned. Otherwise, a circularity has occurred and the result values of the existing entry and
+        // all entries pushed after it are changed to false, and the value false is returned. The target object has
+        // no significance other than to provide a unique identity for a particular type resolution result.
+        function pushTypeResolution(target: Object): boolean {
+            for (let i = 0; i < resolutionCount; i++) {
+                if (resolutionTargets[i] === target) {
+                    do {

```

Comment:
I think it is a little odd to nest these loops. I would move the inner one outside, and put it after the outer one, because it really logically follows the outer one.